vault/logical_system.go

```diff
@@ -214,6 +226,360 @@ type SystemBackend struct {
+		}, nil
+	}
+}
+
+func (b *SystemBackend) handleRaftRemovePeerUpdate() framework.OperationFunc {
+	return func(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+		serverID := d.Get("server_id").(string)
+		if len(serverID) == 0 {
+			return logical.ErrorResponse("no server id provided"), logical.ErrInvalidRequest
+		}
+
+		raftStorage, ok := b.Core.underlyingPhysical.(*raft.RaftBackend)
+		if !ok {
+			return logical.ErrorResponse("raft storage is not in use"), logical.ErrInvalidRequest
+		}
+
+		if err := raftStorage.RemovePeer(ctx, serverID); err != nil {
+			return nil, err
+		}
+		if b.Core.raftFollowerStates != nil {
+			b.Core.raftFollowerStates.delete(serverID)
+		}
+
+		return nil, nil
+	}
+}
+
+func (b *SystemBackend) handleRaftBootstrapChallengeWrite() framework.OperationFunc {
+	return func(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+		serverID := d.Get("server_id").(string)
+		if len(serverID) == 0 {
+			return logical.ErrorResponse("no server id provided"), logical.ErrInvalidRequest
+		}
+
+		// TODO: Make sure the serverID isn't already part of the cluster?
+		uuid, err := uuid.GenerateRandomBytes(16)
+		if err != nil {
+			return nil, err
+		}
+
+		sealAccess := b.Core.seal.GetAccess()
+		eBlob, err := sealAccess.Encrypt(ctx, uuid)
+		if err != nil {
+			return nil, err
+		}
+		protoBlob, err := proto.Marshal(eBlob)
+		if err != nil {
+			return nil, err
+		}
+
+		pendingRaftPeers[serverID] = uuid
+		sealConfig, err := b.Core.seal.BarrierConfig(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		return &logical.Response{
+			Data: map[string]interface{}{
+				"challenge":   base64.StdEncoding.EncodeToString(protoBlob),
+				"seal_config": sealConfig,
+			},
+		}, nil
+	}
+}
+
+func (b *SystemBackend) handleRaftBootstrapAnswerWrite() framework.OperationFunc {
+	return func(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+		raftStorage, ok := b.Core.underlyingPhysical.(*raft.RaftBackend)
+		if !ok {
+			return logical.ErrorResponse("raft storage is not in use"), logical.ErrInvalidRequest
+		}
+
+		serverID := d.Get("server_id").(string)
+		if len(serverID) == 0 {
+			return logical.ErrorResponse("no server_id provided"), logical.ErrInvalidRequest
+		}
+		answerRaw := d.Get("answer").(string)
+		if len(answerRaw) == 0 {
+			return logical.ErrorResponse("no answer provided"), logical.ErrInvalidRequest
+		}
+		clusterAddr := d.Get("cluster_addr").(string)
+		if len(clusterAddr) == 0 {
+			return logical.ErrorResponse("no cluster_addr provided"), logical.ErrInvalidRequest
+		}
+
+		answer, err := base64.StdEncoding.DecodeString(answerRaw)
+		if err != nil {
+			return logical.ErrorResponse("could not base64 decode answer"), logical.ErrInvalidRequest
+		}
+
+		expectedAnswer, ok := pendingRaftPeers[serverID]
+		if !ok {
+			// TODO: Should we return error or hid the fact the peer ID was
```

Comment:
I think either way should have at least a slightly different error so that if it pops up somewhere you know which line it came from.